sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -230,18 +170,45 @@ public void stop() throws IOException {
 
       abstract Builder setRoutingKey(String routingKey);
 
-      abstract Builder setUseCorrelationId(boolean useCorrelationId);
+      abstract Builder setRecordIdPolicy(RecordIdPolicy recordIdPolicy);
 
       abstract Builder setMaxNumRecords(long maxNumRecords);
 
       abstract Builder setMaxReadTime(Duration maxReadTime);
 
+      abstract Builder setTimestampPolicyFactory(TimestampPolicyFactory timestampPolicyFactory);
+
+      abstract Builder setConnectionHandlerProviderFn(
+          SerializableFunction<Void, ConnectionHandler> connectionHandlerProviderFn);
+
       abstract Read build();
     }
 
+    /**
+     * Defines a means of obtaining a connection to RabbitMQ in terms of the provided URI. This url
+     * is expected to contain a UserInfo segment with username and password.
+     *
+     * @param uri URI to use to connect to rabbit
+     * @see ConnectionProviderFromUri for how the URI will be used to obtain a Connection. Use
+     *     {@link #withConnectionHandlerProviderFn(SerializableFunction)} if a different strategy is
+     *     required.
+     */
     public Read withUri(String uri) {
       checkArgument(uri != null, "uri can not be null");
```

Comment:
should we add another check here to make sure that the URI parses?  or do we not care since we're just passing it down anyways (with the assumption that there could possibly be URIs that violate the RFC but the rabbit client still accepts)

Reply:
The general Beam approach seems to be to let it fail at runtime. I think in the case of this Builder pattern it's best validated at either build() time or execution time.